File: src/entities/EntityManager.java
package entities;

import entities.particles.Particle;
import handlers.Camera;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Graphics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;

public class EntityManager {
	private static HashMap<String, AbstractEntity> entities = new HashMap<String, AbstractEntity>();
	
	private static ArrayList<String> entityList = new ArrayList<String>();
	
	public static void add(String name, AbstractEntity entity) {
		entities.put(name, entity);
		entityList.add(name);
	}
	
	public static void remove(String name) {
		entities.remove(name);
		entityList.remove(name);
	}
	
	public static void clear() {
		entities.clear();
		entityList.clear();
	}
	
	public static void updateList() {
		Collections.sort(entityList, new Comparator<String>() {
			@Override
			public int compare(String p1, String p2) {
				double result = entities.get(p1).getDepth() - entities.get(p2).getDepth();
				if (result > 0) {
					return 1;
				}
				if (result < 0) {
					return -1;
				} else {
					return 0;
				}
			}
		});
	}
	
	public static void update(GameContainer gc) {
		for (int i = 0; i < entityList.size(); i++) {
			if (Camera.isEntityInBounds(entities.get(entityList.get(i)))) {
				entities.get(entityList.get(i)).setCollision();
			}
		}
		for (int i = 0; i < entityList.size(); i++) {
			if (entities.get(entityList.get(i)) instanceof Particle || Camera.isEntityInBounds(entities.get(entityList.get(i)))) {
				entities.get(entityList.get(i)).update(gc);
			}
		}
		updateList();
	}
	
	public static void render(GameContainer gc, Graphics g) {
		for (int i = 0; i < entityList.size(); i++) {
			if (Camera.isEntityInBounds(entities.get(entityList.get(i)))) {
				entities.get(entityList.get(i)).render(gc, g);
			}
		}
	}
	
	public static AbstractEntity getEntity(String name) {
		if (entities.containsKey(name)) {
			return entities.get(name);
		} else {
			return null;
		}
	}
}

File: src/entities/EntityManager.kt
package entities

import entities.particles.Particle
import util.Camera
import org.newdawn.slick.GameContainer
import org.newdawn.slick.Graphics
import region.RegionManager
import util.Direction
import util.Direction.*

object EntityManager {
    private val entityMap = EntityMap()

    fun add(particle: Particle) {
        entityMap.add(particle.toString(), particle)
    }

    fun add(name: String, entity: AbstractEntity) {
        entityMap.add(name, entity)
    }

    fun remove(name: String) {
        entityMap.remove(name)
    }

    fun clear() {
        entityMap.clear()
    }

    fun initAdd(name: String, entity: AbstractEntity) {
        add(name, entity)
        entityMap.update()
    }

    fun update(gc: GameContainer) {
        entityMap.entityList
                .filter { Camera.isEntityInBounds(entityMap[it]!!) }
                .forEach { entityMap[it]!!.setCollision() }

        entityMap.entityList
                .filter { entityMap[it] is Particle || Camera.isEntityInBounds(entityMap[it]!!) }
                .forEach { entityMap[it]!!.update(gc) }

        entityMap.update()
    }

    fun render(gc: GameContainer, g: Graphics) {
        for (key in entityMap.entityList) {
            if (Camera.isEntityInBounds(entityMap[key]!!)) {
                entityMap[key]!!.render(gc, g)
            }
        }
    }

    fun getEntity(name: String) = entityMap[name]

    fun areaSwitch(dir: Direction) {
        val yOffset = when (dir) {
            UP -> RegionManager.currentArea.map.height * 16
            DOWN -> -RegionManager.getNeighbor(UP).map.height * 16
            LEFT -> -RegionManager.getNeighbor(RIGHT).getOffset(dir)
            RIGHT -> -RegionManager.getNeighbor(LEFT).getOffset(dir)
        }

        val xOffset = when (dir) {
            LEFT -> RegionManager.getNeighbor(LEFT).map.width * 16
            RIGHT -> -RegionManager.currentArea.map.width * 16
            UP -> -RegionManager.getNeighbor(DOWN).getOffset(dir)
            DOWN -> -RegionManager.getNeighbor(UP).getOffset(dir)
        }

        for (key in entityMap.entityList) {
            val temp = entityMap[key]!!
            temp.setPosition(temp.x + xOffset, temp.y + yOffset)
            temp.updateDepth()
        }
        entityMap.update()
    }
}